### ISSUE TITLE
Use localstack

### ISSUE THREAD

**Issue by gabfssilva** [matsluni/aws-spi-akka-http#1]
Do you think it'd be good to add more tests and use @localstack for the mocks?

**Comment by matsluni** [comment]
Hi @gabfssilva, thank for your contributions and your efforts in Alpakka! 

I am currently on vacation and have to use the mobile to answer, so, please apologize typos and probably shortness in the answer.

What tests would you add? Is anything in particular missing? 

I looked at @localstack when I first thought about how to test this. I settled with other libs at the time because they were easier to add and programmatically to start and shutdown. I haven’t much investigated if it is easy to start and shutdown @localstack mocks.

**Comment by gabfssilva** [comment]
Well, people will be more confortable if the service they're using are also covered by the tests, so, I think the more aws services we can test, the better.

Since the tests require an AWS account, it kind of makes it impossible to automate tests using a CI, for instance, since *every running test* will increase the aws bill, so, that's why I proposed @localstack.

We can run @localstack by running their Docker image and use the endpoints provided to override the endpoint configuration within the tests. I can send a PR with an example.

**Comment by matsluni** [comment]
@gabfssilva, if you want to bring in @localstack, do you know testcontainer? This is basically JUnit integration, see here https://www.testcontainers.org/modules/localstack/. Maybe there is also Scalatest integration. 

This would be good to have to be able to programmatically start and stop the containers.

**Comment by gabfssilva** [comment]
I think it's a good idea. I'll work on this during the week.

**Comment by gabfssilva** [comment]
I see there are two types of tests: the ones using mocks and the other using an actual AWS account.

Do you plan to keep the one using the AWS account and use localstack in the mock tests or do you plan to use only localstack as the "AWS services provider" for the integration tests?

**Comment by matsluni** [comment]
Yes, that is right.

During initial development, I noticed, that not all things had been caught with the mocks. Especially, the signing of requests had errors before. This is why I would keep the tests for now. Also, running these tests infrequently does not create costs for me.

Do you know if localstack is able to check signed requests (with artificial credentials)?

**Comment by gabfssilva** [comment]
I don't think so, since Localstack also uses services like ElasticMQ behind the scenes, it can suppress a few errors.

That's a good point, I'll keep both tests and change the mocked one to run with Localstack.

**Comment by gabfssilva** [comment]
Just to update: I'm experiencing an issue using s3 from localstack with this sdk. Don't know why, it keeps throwing this exception: `SdkClientException: Data read has a different checksum than expected.`

I don't think this is caused by this library, since the tests currently work using s3mock and also aws s3, but, we need to check why this is happening.

**Comment by matsluni** [comment]
Thanks for the update. 
I remember I had also some issues with some version in the past, but I thought that this is fixed by now. I think this was the issue back than: https://github.com/aws/aws-sdk-java-v2/issues/965.

I also found this, this looks like the bug you are seeing: https://github.com/aws/aws-sdk-java-v2/issues/953

Back then I enabled this, to temporarily disable checksum check: https://github.com/matsluni/aws-spi-akka-http/blob/db285a36bb92955993ef48cb6747231e729d3569/src/it/scala/com/github/matsluni/akkahttpspi/s3/ITTestS3.scala#L37

I hope this helps.

**Comment by gabfssilva** [comment]
I have bad news and good news.

The bad news: it seems that the s3 tests don't work very well with localstack
The good news: the same issue seems to happen with the Netty connector, so, this lib is not the problem. =P

We could use other mocks and use them as containers, but, I think it'd be a cumbersome thing to do.

What do you think? Remove the s3 test from the mock tests and add others or use s3mock as a docker container to test s3?

**Comment by matsluni** [comment]
> I have bad news and good news.

😄 

Ok, as far as I understand you, you have some aws service test covered with localstack (sqs, dns, lambda, dynamodb?), but s3 is not working, right?

What about you open a PR with those changes, but keep s3 tests untouched. There are plenty of other s3 mocks. For me it would also be okay to have a mixed setup of mock tests, some with localstack (when there is viable jvm option available) and others without it (when there is a viable jvm option available, see dynamodb tests). 

For S3, I would look into other mock services and change the from current one to something else, also to unblock #3.

What do you think?

**Comment by gabfssilva** [comment]
I tried a few s3 mocks, but, none of them work. Some work with the netty connectors, others don't.

The issue is: the upload do not work. The server says it has uploaded, but, if you download the file, it's empty.

The [adobe/S3Mock](https://github.com/adobe/S3Mock) seems to work quite well with the netty connector, maybe we should stick to it to make aws-spi-akka-http work.

I also tried to upload something to adobe/S3Mock using Alpakka S3 connector (which uses a completely different implementation) and the same happens: if you try to download, the file is empty.

Maybe this is a issue with these mocks and Akka HTTP? Don't know.

Let's keep the s3 tests and see why the mocks don't work well with Akka HTTP.

I'll send a PR today.

**Comment by gabfssilva** [comment]
This is funny: I didn't notice the Java test that used findify/s3mock, so I decided to give it another try. 

It works perfectly using directly the jar, but, somehow when I run the same test using the docker container, it don't work and I get an ` java.util.concurrent.CompletionException: software.amazon.awssdk.services.s3.model.S3Exception:  (Service: S3, Status Code: 500, Request ID: null)`
I also noticed the last docker tag is `0.2.4`, while the jar we were using was `0.2.5`, so I thought that it may be it, but, I changed the jar version to 0.2.4 and it also didn't work.

Maybe the docker image is being badly created, don't know.
